### PR TITLE
Compile Concurrently at Synthesis Boundaries

### PR DESCRIPTION
**Pull request by alex-mckenna**
This PR starts the work of making Clash compile concurrently, by generating HDL for different entities asynchronously. For designs with many entities, this can already provide a fairly dramatic improvement as seen in the new `clash-benchmark-concurrency` suite:

Before:

```
ManyEntitiesEqual
time                 27.66 s    (27.43 s .. 28.11 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 27.39 s    (27.29 s .. 27.48 s)
std dev              144.0 ms   (75.86 ms .. 203.0 ms)
variance introduced by outliers: 19% (moderately inflated)

ManyEntitiesVaried
time                 14.56 s    (14.18 s .. 14.97 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.72 s    (14.64 s .. 14.80 s)
std dev              107.4 ms   (2.795 ms .. 130.4 ms)
variance introduced by outliers: 19% (moderately inflated)
```

After:

```
ManyEntitiesEqual
time                 4.215 s    (4.033 s .. 4.389 s)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 4.195 s    (4.152 s .. 4.219 s)
std dev              44.01 ms   (6.522 ms .. 56.38 ms)
variance introduced by outliers: 19% (moderately inflated)

ManyEntitiesVaried
time                 7.790 s    (7.172 s .. 8.436 s)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 7.603 s    (7.475 s .. 7.771 s)
std dev              163.1 ms   (27.67 ms .. 208.1 ms)
variance introduced by outliers: 19% (moderately inflated)
```

## Still TODO:

  - [x] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files
  - [x] ~Change the message when ignoring a cache to mention the entity~ Only print the message when ignoring a cache once

[comment]: # (Depending on the change, some of these points may not be necessary. If so, leave the boxes unchecked so it is easier for a reviewer to see what has been omitted.)

### CODE REVIEW

**Comment by martijnbastiaan** [comment]
Related issue: https://github.com/clash-lang/clash-compiler/issues/354

**Comment by alex-mckenna** [comment]
@martijnbastiaan Good catch!

File: clash-lib/src/Clash/Driver.hs [hunk 1]
```diff
@@ -287,109 +323,99 @@ generateHDL
...
+      (modNameS, fmap Data.Text.pack -> prefixM) = prefixModuleName (hdlKind (undefined :: backend)) (opt_componentPrefix opts) annM modName1
       modNameT  = Data.Text.pack modNameS
       iw        = opt_intWidth opts
-      hdlsyn    = opt_hdlSyn opts
       forceUnd  = opt_forceUndefined opts
       xOpt      = coerce (opt_aggressiveXOptBB opts)
       enums     = coerce (opt_renderEnums opts)
       hdlState' = setModName modNameT
-                $ fromMaybe (initBackend iw hdlsyn escpIds lwIds forceUnd xOpt enums :: backend) hdlState
+                  -- TODO initBackend should just take ClashOpts as an argument.
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Driver.hs:383]: :heavy_check_mark:

File: clash-lib/src/Clash/Driver.hs [hunk 2]
```diff
@@ -411,18 +438,21 @@ generateHDL reprs domainConfs bindingsMap hdlState primMap tcm tupTcm typeTrans
       putStrLn $ "Clash: Normalization took " ++ prepNormDiff
 
       -- 3. Generate netlist for topEntity
-      (topComponent, netlist, seen2) <-
-        genNetlist isTb opts reprs transformedBindings topEntityMap compNames primMap
-                   tcm typeTrans iw ite (SomeBackend hdlState') seen0 hdlDir prefixM topEntity
+      (topComponent, netlist) <- modifyMVar seenV $ \seen -> do
+        (topComponent, netlist, seen') <-
+          -- TODO My word, this has far too many arguments.
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Driver.hs:443]: :heavy_check_mark:

File: clash-lib/clash-lib.cabal [hunk 3]
```diff
@@ -130,6 +130,7 @@ Library
                       aeson-pretty            >= 0.8      && < 0.9,
                       ansi-terminal           >= 0.8.0.0  && < 0.12,
                       array,
+                      async,
```
Comment (martijnbastiaan) [inline clash-lib/clash-lib.cabal:133]: Constrain it to `<= 2.3` to prevent breakages on Hackage

**Review by martijnbastiaan — APPROVED** [review]
Nice work

File: clash-lib/src/Clash/Driver.hs (hunk 1 above)
Comment (alex-mckenna) [reply clash-lib/src/Clash/Driver.hs:383]: Related: #1932

File: clash-lib/src/Clash/Driver.hs (hunk 2 above)
Comment (alex-mckenna) [reply clash-lib/src/Clash/Driver.hs:443]: Related: #1932

**Pull request merged**